lib/phoenix/pubsub.ex

```diff
@@ -21,150 +20,39 @@ defmodule Phoenix.PubSub do
-  end
+  def subscribers(server, topic),
+    do: call(server, {:subscribers, topic})
 
   @doc """
-  Adds subsriber pid to the given topic
-
-    * pid - The Pid of the subscriber
-    * topic_name - The String name of the topic
-
-  ## Examples
-
-      iex> PubSub.subscribe(self, "mytopic")
-
+  Broadcasts message on given topic
   """
-  def subscribe(pid, topic_name) do
-    :ok = create(topic_name)
-    call {:subscribe, pid, group(topic_name)}
-  end
+  def broadcast(server, topic, message),
+    do: call(server, {:broadcast, :none, topic, message})
 
   @doc """
-  Removes the given subscriber from the topic
-
-    * pid - The Pid of the subscriber
-    * topic_name - The String name of the topic
-
-  ## Examples
-
-      iex> PubSub.unsubscribe(self, "mytopic")
-
+  Broadcasts message to all but sender on given topic
   """
-  def unsubscribe(pid, topic_name) do
-    call {:unsubscribe, pid, group(topic_name)}
-  end
+  def broadcast_from(server, from_pid, topic, message),
+    do: call(server, {:broadcast, from_pid, topic, message})
 
   @doc """
-  Returns the List of subsriber pids for the give topic
-
-  ## Examples
-
-      iex> PubSub.subscribers("mytopic")
-      []
-      iex> PubSub.subscribe(self, "mytopic")
-      :ok
-      iex> PubSub.subscribers("mytopic")
-      [#PID<0.41.0>]
-
+  Returns lists of strings of all topics under pg2
   """
-  def subscribers(topic_name) do
-    case :pg2.get_members(group(topic_name)) do
-      {:error, {:no_such_group, _}} -> []
-      members -> members
-    end
-  end
-
-  @doc """
-  Broadcasts a message to the topic's subscribers
-
-    * topic_name - The String name of the topic
-    * message - The term to broadcast
-
-  ## Examples
-
-      iex> PubSub.broadcast("mytopic", :hello)
-
-  To exclude the broadcaster from receiving the message, use `broadcast_from/3`
-  """
-  def broadcast(topic_name, message) do
-    broadcast_from(:global, topic_name, message)
-  end
-
-  @doc """
-  Broadcasts a message to the topics's subscribers, excluding
-  broadcaster from receiving the message it sent out
-
-    * topic_name - The String name of the topic
-    * message - The term to broadcast
-
-  ## Examples
-
-      iex> PubSub.broadcast_from(self, "mytopic", :hello)
-
-  """
-  def broadcast_from(from_pid, topic_name, message) do
-    topic_name
-    |> subscribers
-    |> Enum.each fn
-      pid when pid != from_pid -> send(pid, message)
-      _pid -> :ok
-    end
-  end
-
-  @doc """
-  Check if PubSub is active. To be active it must be created and have subscribers
-  """
-  def active?(topic_name) do
-    call {:active?, group(topic_name)}
-  end
-
-  @doc """
-  Returns a List of all Phoenix PubSubs from :pg2
-  """
-  def list do
-    :pg2.which_groups |> Enum.filter(&match?({@pg_prefix, _}, &1))
-  end
-
-  defp call(message), do: GenServer.call(Server.leader_pid, message)
-
-  defp group(topic_name), do: {@pg_prefix, topic_name}
+  def list(server), do: call(server, :list)
```

Comment:
Option 2 definitely. I didn't think about how easy it would be with the named process